monitoring/monitoring.go

```diff
@@ -0,0 +1,58 @@
+package monitoring
+
+import (
+	"google.golang.org/grpc/codes"
+)
+
+type RpcType string
+
+const (
+	Unary     RpcType = "unary"
+	Streaming RpcType = "streaming"
+)
+
+// RpcMonitor is a per-RPC datastructure.
+type RpcMonitor interface {
+	// ReceivedMessage is called on every stream message received by the monitor.
+	ReceivedMessage()
+
+	// SentMessage is called on every stream message sent by the monitor.
+	SentMessage()
+
+	// Handled is called whenever the RPC handling completes (with OK or AppError).
+	Handled(code codes.Code)
+
+	// Erred is called whenever the RPC failed due to RPC-layer errors.
+	Erred(err error)
+}
+
+// ServerMonitor allocates new per-RPC monitors on the server side.
+type ServerMonitor interface {
+	// NewMonitor allocates a new per-RPC monitor, also signifying the start of an RPC call.
+	NewServerMonitor(rpcType RpcType, fullMethod string) RpcMonitor
```

Comment:
But then they would clash with `ClientMonitor`. Or would you prefer me to collapse them into one?